uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/token/JwtBearerGrantMockMvcTests.java

```diff
@@ -84,6 +89,85 @@ void default_zone_jwt_grant() throws Exception {
             .andExpect(jsonPath("$.access_token").isNotEmpty());
     }
 
+    @Test
+    void non_default_zone_jwt_grant_user_update() throws Exception {
+        BaseClientDetails targetZoneClient = new BaseClientDetails(generator.generate(), "", "openid", "password", null);
+        targetZoneClient.setClientSecret(SECRET);
+        String subdomain = generator.generate().toLowerCase();
+        IdentityZone targetZone = MockMvcUtils.createOtherIdentityZoneAndReturnResult(subdomain,
+                mockMvc,
+                webApplicationContext,
+                targetZoneClient,
+                false, IdentityZoneHolder.getCurrentZoneId()).getIdentityZone();
+        ScimUser targetZoneUser = createUser(targetZone);
+
+        String originZoneOriginKey = createProvider(targetZone, getTokenVerificationKey(originZone.getIdentityZone()));
+
+        //Check for internal User
+        String targetZoneIdToken = getUaaIdToken(targetZone, targetZoneClient, targetZoneUser);
+        String accessTokenForTargetZoneUser = performJWTBearerGrantForJWT(targetZone, targetZoneIdToken);
+
+        //Verify JWT Bearer did not change values of internal User
+        ScimUser targetUserAfterGrant = getScimUser(targetZoneUser.getUserName(), OriginKeys.UAA, targetZone.getId());
+        assertEquals(targetZoneUser.getUserName(), targetUserAfterGrant.getUserName());
+        assertEquals(targetZoneUser.getExternalId(), targetUserAfterGrant.getExternalId());
+
+        //Check for user of registered IdP
+        String originZoneIdToken = getUaaIdToken(originZone.getIdentityZone(), originClient, originUser);
+        String accessTokenForOriginZoneUser = performJWTBearerGrantForJWT(targetZone, originZoneIdToken);
+        Map<String, Object> originUserClaims = JwtTokenUtils.getClaimsForToken(accessTokenForOriginZoneUser);
+
+        //Verify values for new shadow user set
+        ScimUser originShadowUser = getScimUser(originUser.getEmails().get(0).getValue(), originZoneOriginKey, targetZone.getId());
+        assertEquals(originShadowUser.getUserName(), originUserClaims.get("user_name"));
+        assertEquals(originShadowUser.getExternalId(), originUser.getId());
+
+        //JWT Bearer with token from target Zone and external User
+        performJWTBearerGrantForJWT(targetZone, accessTokenForOriginZoneUser);
+
+        //Verify username and External ID not changed after this internal grant
+        ScimUser originShadowUserAfterExchange = getScimUser(originUser.getEmails().get(0).getValue(), originZoneOriginKey, targetZone.getId());
+        assertEquals(originShadowUser.getUserName(), originShadowUserAfterExchange.getUserName());
+        assertEquals(originShadowUser.getExternalId(), originShadowUserAfterExchange.getExternalId());
+    }
+
+    @Test
+    void default_zone_jwt_grant_user_update_same_zone_with_registration() throws Exception {
```

Comment:
If I'm reading this correctly, "same_zone_with_registration" means "registering a zone as an External IDP in itself", which seems like an unusual use case. Is this a documented use case? If not, is there a reason why we should make sure this use case works?

Isn't the JWT bearer grant that involves an external IDP already tested elsewhere?

Reply:
I agree that this is a rather unusual case - but it is handled when resolving the internal idp at the moment and also used in other tests, so I wanted to handle this case here too.

Reply:
Could you help me find where those other tests are? I wanna try to understand why they are written that way.

Reply:
Hi,

I know that at least the tests the [PasswordGrantIT](https://github.com/cloudfoundry/uaa/blob/develop/uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/PasswordGrantIT.java) that is using it this way (though only for the OIDC password grant). Not sure if there are others also doing it this way.

Reply:
Interesting, I wonder if this is for the convenience of test setup or a real-life use case. But the current code to support this use case, though confusing to read (mitigated via a comment), likely won't cause any issue, so let's leave it be.